tests/components/knx/conftest.py

```diff
@@ -1,15 +1,128 @@
-"""conftest for knx."""
+"""Conftest for the KNX integration."""
+from __future__ import annotations
 
-from unittest.mock import AsyncMock, Mock
+import asyncio
+from unittest.mock import AsyncMock, Mock, patch
 
 import pytest
+from xknx import XKNX
+from xknx.telegram.apci import APCI, GroupValueRead, GroupValueResponse, GroupValueWrite
 
+from homeassistant.components.knx.const import DOMAIN as KNX_DOMAIN
+from homeassistant.core import HomeAssistant
+from homeassistant.setup import async_setup_component
 
-@pytest.fixture(autouse=True)
-def knx_ip_interface_mock():
-    """Create a knx ip interface mock."""
-    mock = Mock()
-    mock.start = AsyncMock()
-    mock.stop = AsyncMock()
-    mock.send_telegram = AsyncMock()
-    return mock
+
+class KNXTestKit:
+    """Test helper for the KNX integration."""
+
+    def __init__(self, hass: HomeAssistant):
+        """Init KNX test helper class."""
+        self.hass: HomeAssistant = hass
+        self.xknx: XKNX
+        # outgoing telegrams will be put in the Queue instead of sent to the interface
+        # telegrams to an InternalGroupAddress won't be queued here
+        self._outgoing_telegrams: asyncio.Queue = asyncio.Queue()
+
+    async def setup_integration(self, config):
+        """Create the KNX integration."""
+
+        def knx_ip_interface_mock():
+            """Create a xknx knx ip interface mock."""
+            mock = Mock()
+            mock.start = AsyncMock()
+            mock.stop = AsyncMock()
+            mock.send_telegram = AsyncMock(side_effect=self._outgoing_telegrams.put)
+            return mock
+
+        with patch("xknx.xknx.KNXIPInterface", return_value=knx_ip_interface_mock()):
+            await async_setup_component(self.hass, KNX_DOMAIN, {KNX_DOMAIN: config})
+            await self.hass.async_block_till_done()
+            self.xknx = self.hass.data[KNX_DOMAIN].xknx
```

Comment:
Please don't access integration details like `hass.data` in the tests. Avoiding that will make the tests more robust.

Reply:
🤔 I'll need some reference to the KNX library (xknx) to be able to inject (fake incoming) telegrams later when platforms are to be tested to trigger the `Entity.async_write_ha_state()` callback (local push). 
Previously it was extracted from a Mock which does the same but is harder to read and seems kind of hacky to me. https://github.com/home-assistant/core/blob/2021.7.3/tests/components/knx/__init__.py#L14-L19
Am I missing the obvious way to do that?

Reply:
Patching the library and accessing the mock is the way to go.